### PR TITLE
Improvements to Shared Heartbeat Driver

### PR DESCRIPTION
**Pull request by liam-ilan**
### Summary
This pull will allow us to see which board's heartbeat fails, when a dependent board faults due to a missing heartbeat, and also includes some refactoring to replace all one hot encodings with arrays. The heartbeat monitor now manages CAN on it's own, by being passed function pointers to CAN methods on creation.

### Changelist
- Heartbeat `status` is now stores the status of individual boards in an array.
- **Rework**: `App_SharedHeartbeatMonitor_Create`: now gets CAN heartbeat getters, updaters, fault setters, and board setter callbacks, since the heartbeat monitor now manages it's own CAN.
- **Rework**: `App_SharedHeartbeatMonitor_CheckIn`: this method now handles the whole check in process.
- **New method**: `App_SharedHeartbeatMonitor_BroadcastFaults`: handles fault broadcasting automatically.
- **New method**: `App_SharedHeartbeatMonitor_CheckFaults`: checks heartbeat faults a given board relies on.
- New CAN Messages/Signals
  - Every board now has additional faults in the CAN table, corresponding to the boards they rely on.
- Heartbeat Monitor now has Unit Tests

### Testing Done
- [x] Passing CI tests
- [x] Flashed on car

### Resolved Issues
None.

### Checklist
*Please change `[ ]` to `[x]` when you are ready.*
- [x] I have read and followed the code conventions detailed in [README.md](../README.md) (*This will save time for both you and the reviewer!*).
- [x] If this pull request is longer then **500** lines, I have provided *explicit* justification in the summary above explaining why I *cannot* break this up into multiple pull requests (*Small PR's are faster and less painful for everyone involved!*).

*This PR is greater than 500 lines... but encapsulates two features better kept together (heartbeat rework and specific heartbeat faults for each board)*

### CODE REVIEW

**Comment by liam-ilan** [comment]
Status: working on debugging FSM failing tests...
Locally:
```
[  FAILED  ] 5 tests, listed below:
[  FAILED  ] FsmFaultTest.check_state_transition_fault_state_heartbeat_timeout
[  FAILED  ] FsmFaultTest.papps_ocsc_sets_mapped_pedal_percentage_to_zero_and_sets_fault
[  FAILED  ] FsmFaultTest.sapps_ocsc_sets_mapped_pedal_percentage_to_zero_and_sets_fault
[  FAILED  ] FsmFaultTest.apps_disagreement_sets_mapped_pedal_percentage_to_zero_and_sets_fault
[  FAILED  ] FsmFaultTest.brake_actuated_sets_mapped_pedal_percentage_to_zero_and_sets_fault_if_papps_pressed
```

**Comment by liam-ilan** [comment]
Current status:
- FSM tests fail... no idea why.
- CubeMX Code Gen for PDM failing... also no idea why.

**Comment by liam-ilan** [comment]
Status: FSM tests still failing
- FSM is broadcasting a heartbeat fault when there should be none

**Comment by liam-ilan** [comment]
Status:
Found error. Heartbeat status is larger than 5 bits, so when we check for equality with `0x1F` (good status), the check fails (`0x1F` assumes a 5 bit status size).

**Comment by liam-ilan** [comment]
### Status:
- All CI tests pass.
- More extensive car testing to be done when battery is back in car.

#### TODO
- More extensive car testing [ ]
- Remove HeartbeatStatus CAN message, replace with CAN fault for every board [ ]

**Comment by liam-ilan** [comment]
# Status
Feature should work, tests need to be fixed to use new CAN alerts.

## TODO

- [ ] Update tests to use new CAN alerts/warnings.
- [ ] Move configs for can funcs to main with `extern`
- [ ] Test on car

**Comment by liam-ilan** [comment]
# Status
- All CI tests passing.
- Car testing is next.
- Ready for code review.

File: can_bus/json/FSM/FSM_tx.json [hunk 1]
```diff
@@ -1,181 +1,181 @@
...
-			"State": {
-				"enum": "State_Machine"
-			}
-		}
-	},
-	"AlertsContext": {
-		"msg_id": 301,
-		"cycle_time": 1000,
-		"signals": {
-			"WatchdogTimeoutTaskName": {
```
Comment (gtaharaedmonds) [inline can_bus/json/FSM/FSM_tx.json:19]: What's with all the whitespace diffs?

File: can_bus/json/PDM/PDM_alerts.json [hunk 2]
```diff
@@ -11,7 +11,8 @@
         "StackWaterMarkHighTaskCanTx",
         "WatchdogTimeout",
         "TxOverflow",
-        "RxOverflow"
+        "RxOverflow",
+        "MissingBMSHeartbeat"
```
Comment (gtaharaedmonds) [inline can_bus/json/PDM/PDM_alerts.json:15]: The PDM now relies on the BMS for some signals -> You could make this a fault now

File: firmware/shared/src/app/App_SharedHeartbeatMonitor.c [hunk 3]
```diff
@@ -26,7 +17,7 @@ struct HeartbeatMonitor *App_SharedHeartbeatMonitor_Create(
     heartbeat_monitor->previous_timeout_ms   = 0U;
     heartbeat_monitor->heartbeats_checked_in = 0;
     heartbeat_monitor->heartbeats_to_check   = heartbeats_to_check;
-    heartbeat_monitor->status                = true;
+    heartbeat_monitor->status                = 0x1F; // mark all heartbeats good
```
Comment (gtaharaedmonds) [inline firmware/shared/src/app/App_SharedHeartbeatMonitor.c:20]: Where does this `0x1F` magic number come from?

File: firmware/shared/src/app/App_SharedHeartbeatMonitor.c [hunk 4]
```diff
@@ -45,7 +36,8 @@ bool App_SharedHeartbeatMonitor_Tick(struct HeartbeatMonitor *const heartbeat_mo
         heartbeat_monitor->previous_timeout_ms = current_ms;
 
         // Check if the board received all the heartbeats it's listening for
-        heartbeat_monitor->status = heartbeat_monitor->heartbeats_to_check == heartbeat_monitor->heartbeats_checked_in;
+        heartbeat_monitor->status =
+            ~(heartbeat_monitor->heartbeats_to_check) | (heartbeat_monitor->heartbeats_checked_in);
```
Comment (gtaharaedmonds) [inline firmware/shared/src/app/App_SharedHeartbeatMonitor.c:40]: Could add some more documentation? Not super clear how this works just by looking at it.
Also consider keeping track of heartbeat statuses as an array if its more clear - whatever you think is best.

File: firmware/shared/src/app/App_SharedHeartbeatMonitor.c [hunk 5]
```diff
@@ -60,3 +52,63 @@ void App_SharedHeartbeatMonitor_CheckIn(
...
+    for (int i = 0; i < NUM_HEARTBEAT_INDICIES; i++)
+    {
+        // build message to broadcast
+        int  mask = (1 << i);
+        bool res  = (mask & state) >> i || !req;
+
+        // if a callback is provided, send a message through
+        // broadcast !res, as faults are 1 for fail, 0 for pass
+        if (callbackList[i] != NULL)
+            (*(callbackList[i]))(!res);
```
Comment (gtaharaedmonds) [inline firmware/shared/src/app/App_SharedHeartbeatMonitor.c:78]: nit: we use curly braces, even for just one liners

File: firmware/shared/src/app/App_SharedHeartbeatMonitor.c [hunk 6]
```diff
@@ -60,3 +52,63 @@ void App_SharedHeartbeatMonitor_CheckIn(
...
+{
+    // get callbacks and throw in list for itteration
+    bool (*callbackList[NUM_HEARTBEAT_INDICIES])() = { callbacks.bms, callbacks.dcm, callbacks.pdm, callbacks.fsm,
+                                                       callbacks.dim };
+
+    // check all callbacks return false
+    bool res = true;
+    for (int i = 0; i < NUM_HEARTBEAT_INDICIES; i++)
+        if (callbackList[i] != NULL)
+            res &= !(*(callbackList[i]))();
```
Comment (gtaharaedmonds) [inline firmware/shared/src/app/App_SharedHeartbeatMonitor.c:95]: nit again: curly braces

File: firmware/thruna/BMS/Src/App/states/App_AllStates.c [hunk 7]
```diff
@@ -40,7 +48,10 @@ static bool App_SendAndReceiveHeartbeat(struct HeartbeatMonitor *hb_monitor)
         App_CanRx_PDM_Heartbeat_Update(false);
     }
 
-    const bool missing_hb = !App_SharedHeartbeatMonitor_Tick(hb_monitor);
+    const enum HeartbeatOneHot hb_state = App_SharedHeartbeatMonitor_Tick(hb_monitor);
+    App_SharedHeartbeatMonitor_BroadcastFaults(hb_state, heartbeatCanSetters, true);
+
+    const bool missing_hb = !App_SharedHeartbeatMonitor_ValidateState(hb_state);
```
Comment (gtaharaedmonds) [inline firmware/thruna/BMS/Src/App/states/App_AllStates.c:54]: Hmm I was hoping that the previous lines where we get the heartbeats and check them in could be handled by the heartbeat driver itself
And then we only have to call the tick function from the state machine

File: can_bus/json/FSM/FSM_tx.json (hunk 1 above)
Comment (liam-ilan) [reply can_bus/json/FSM/FSM_tx.json:19]: There was some inconsistent indentation before hand... I can undo the whitespace diff, keep it, or merge it in a new PR... LMK what you think...

**Comment by liam-ilan** [comment]
> 2. consider switching to arrays to keep track of state, bitmask stuff is confusing and not necessary for this application

Agreed! Though I think it would make sense to have a heartbeat flag struct... just because each status bit inherently has a "name" (the board to which it corresponds to)... I'll work on dropping the one-hot encoding in favor of this.

File: .vscode/settings.json [hunk 8]
```diff
@@ -9,5 +9,8 @@
 	],
 	"cmake.buildDirectory": "${workspaceFolder}/${env:BUILD_DIR}",
 	"cmake.configureOnEdit": false,
-	"cmake.configureOnOpen": true,
+  "cmake.configureOnOpen": true,
+  "files.associations": {
+    "app_sharedheartbeatmonitor.h": "c"
+  },
```
Comment (gtaharaedmonds) [inline .vscode/settings.json:15]: revert these changes please

File: firmware/thruna/BMS/Src/main.c [hunk 9]
```diff
@@ -180,6 +180,12 @@ struct PrechargeRelay *  precharge_relay;
 struct TractiveSystem *  ts;
 struct Clock *           clock;
 struct Eeprom *          eeprom;
+// BMS checks DCM, PDM, FSM
+bool heartbeat_monitor_checklist[HEARTBEAT_BOARD_COUNT] = { [BMS_HEARTBEAT_BOARD] = false,
+                                                            [DCM_HEARTBEAT_BOARD] = true,
+                                                            [PDM_HEARTBEAT_BOARD] = true,
+                                                            [FSM_HEARTBEAT_BOARD] = true,
+                                                            [DIM_HEARTBEAT_BOARD] = false };
```
Comment (gtaharaedmonds) [inline firmware/thruna/BMS/Src/main.c:188]: Right now we're providing half of the config info in `main.c`, and the other half the in the state machine all states function. 
I don't like this because its messy, and also makes it easier to make a mistake where you provide mistmatched configs. Also also, now you need to define the setters/getters wherever you use the heartbeat driver (what if you wanted to call its functions from a different file?).
Let me know what you think, but I feel like refactoring it so you provide all configs here would be a nice change, and also simplify the state machine code a lot.

**Review by gtaharaedmonds — CHANGES_REQUESTED** [review]
overall looks great, left one comment about how config info is provided that would be a medium sized change, let me know what you think of that. tests are awesome.

File: .vscode/settings.json (hunk 8 above)
Comment (liam-ilan) [reply .vscode/settings.json]: got it!

File: firmware/thruna/BMS/Src/main.c (hunk 9 above)
Comment (liam-ilan) [reply firmware/thruna/BMS/Src/main.c]: All heartbeat can methods now live in the heartbeat monitor itself, and get set when the monitor is created. All the config info is now in main.c as requested! :D

File: .vscode/settings.json [hunk 10]
```diff
@@ -8,5 +8,5 @@
 		}
 	],
 	"cmake.configureOnEdit": false,
-	"cmake.configureOnOpen": false,
+	"cmake.configureOnOpen": true,
```
Comment (gtaharaedmonds) [inline .vscode/settings.json:11]: can you revert this

File: firmware/thruna/PDM/src/cubemx/Src/main.c [hunk 11]
```diff
@@ -249,6 +250,37 @@ static const EfuseConfig efuse_configs[NUM_EFUSE_CHANNELS] = {
     }
 };
 
+// config to forward can functions to shared heartbeat
+// PDM rellies on BMS
+bool heartbeatMonitorChecklist[HEARTBEAT_BOARD_COUNT] = { [BMS_HEARTBEAT_BOARD] = true,
+                                                          [DCM_HEARTBEAT_BOARD] = false,
+                                                          [PDM_HEARTBEAT_BOARD] = false,
+                                                          [FSM_HEARTBEAT_BOARD] = true,
+                                                          [DIM_HEARTBEAT_BOARD] = true };
```
Comment (gtaharaedmonds) [inline firmware/thruna/PDM/src/cubemx/Src/main.c:259]: PDM does not rely on FSM or DIM

File: firmware/thruna/DIM/src/cubemx/Src/main.c [hunk 12]
```diff
@@ -288,6 +288,37 @@ static const GlobalsConfig globals_config = {
     .dim_status_led = &dim_status_led,
 };
 
+// config to forward can functions to shared heartbeat
+// DIM rellies on BMS
+bool heartbeatMonitorChecklist[HEARTBEAT_BOARD_COUNT] = { [BMS_HEARTBEAT_BOARD] = true,
+                                                          [DCM_HEARTBEAT_BOARD] = false,
+                                                          [PDM_HEARTBEAT_BOARD] = false,
+                                                          [FSM_HEARTBEAT_BOARD] = false,
+                                                          [DIM_HEARTBEAT_BOARD] = false };
```
Comment (gtaharaedmonds) [inline firmware/thruna/DIM/src/cubemx/Src/main.c:297]: DIM should rely on every board, since it displays the faults/warning of every board

**Review by gtaharaedmonds — CHANGES_REQUESTED** [review]
looks great, a couple updates we should make to the board dependency graph, but otherwise LGTM!!!

**Review by gtaharaedmonds — APPROVED** [review]
LGTM!
